File: src/main/java/openmedstack/events/EventHandlerBase.java
package openmedstack.events;

import openmedstack.MessageHeaders;

import java.util.concurrent.CompletableFuture;

public abstract class EventHandlerBase<T extends BaseEvent> implements IHandleEvents<T> {

    public Boolean canHandle(Class type) {
        return BaseEvent.class.isAssignableFrom(type);
    }

    public CompletableFuture handle(T evt, MessageHeaders headers) {
        return verifyUserToken(headers.getUserToken()).thenApply(b -> {
            if (b) {
                return handleInternal(evt, headers);
            }else{
                return new CompletableFuture();
            }
        });
    }

    protected CompletableFuture<Boolean> verifyUserToken(String token) {
        return new CompletableFuture<Boolean>().completeAsync(() -> Boolean.TRUE);
    }

    protected abstract CompletableFuture handleInternal(T evt, MessageHeaders headers);
}

File: src/main/kotlin/org/openmedstack/events/EventHandlerBase.kt
package org.openmedstack.events

import org.openmedstack.MessageHeaders
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.concurrent.CompletableFuture
import kotlin.reflect.KClass

abstract class EventHandlerBase<T : BaseEvent?> : IHandleEvents<T> {
    final override fun canHandle(type: Class<*>): Boolean {
        return Class.forName(javaClass.typeParameters[0].name).isAssignableFrom(type)
    }

    override fun handle(evt: T, headers: MessageHeaders): CompletableFuture<*> {
        return verifyUserToken(headers.userToken).thenApply { b: Boolean ->
            if (b) {
                return@thenApply handleInternal(evt, headers)
            } else {
                return@thenApply CompletableFuture<Any>()
            }
        }
    }

    protected fun verifyUserToken(token: String?): CompletableFuture<Boolean> {
        return CompletableFuture<Boolean>().completeAsync { java.lang.Boolean.TRUE }
    }

    protected abstract fun handleInternal(evt: T, headers: MessageHeaders?): CompletableFuture<*>?
}
